File: packages/app-mobile/ios/AppMobileModule.swift
import ExpoModulesCore

public class AppMobileModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AppMobile")

    Events("onTick")

    Function("add") { (left: Int, right: Int) -> Int in
      Int(add(left: UInt64(left), right: UInt64(right)))
    }

    Function("subtract") { (left: Int, right: Int) -> Int in
      Int(subtract(left: UInt64(left), right: UInt64(right)))
    }

    Function("multiply") { (left: Int, right: Int) -> Int in
      Int(multiply(left: UInt64(left), right: UInt64(right)))
    }

    Function("divide") { (left: Int, right: Int) -> Int? in
      divide(left: UInt64(left), right: UInt64(right)).map { Int($0) }
    }

    AsyncFunction("delayedAdd") { (left: Int, right: Int, delayMs: Int) -> Int in
      Int(await delayedAdd(left: UInt64(left), right: UInt64(right), delayMs: UInt64(delayMs)))
    }

    Function("startBackgroundTask") {
      startBackgroundTask(callback: AppMobileBackgroundCallback(module: self))
    }

    Function("stopBackgroundTask") {
      stopBackgroundTask()
    }
  }
}

class AppMobileBackgroundCallback: BackgroundTaskCallback {
  private weak var module: AppMobileModule?

  init(module: AppMobileModule) {
    self.module = module
  }

  func onTick(count: UInt64) {
    module?.sendEvent("onTick", ["count": Int(count)])
  }
}

File: packages/app-mobile/android/src/main/java/expo/modules/appmobile/AppMobileModule.kt
package expo.modules.appmobile

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.app_mobile.BackgroundTaskCallback
import uniffi.app_mobile.add
import uniffi.app_mobile.delayedAdd
import uniffi.app_mobile.divide
import uniffi.app_mobile.multiply
import uniffi.app_mobile.startBackgroundTask
import uniffi.app_mobile.stopBackgroundTask
import uniffi.app_mobile.subtract

class AppMobileModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AppMobile")

    Events("onTick")

    Function("add") { left: Long, right: Long ->
      add(left.toULong(), right.toULong()).toLong()
    }

    Function("subtract") { left: Long, right: Long ->
      subtract(left.toULong(), right.toULong()).toLong()
    }

    Function("multiply") { left: Long, right: Long ->
      multiply(left.toULong(), right.toULong()).toLong()
    }

    Function("divide") { left: Long, right: Long ->
      divide(left.toULong(), right.toULong())?.toLong()
    }

    AsyncFunction("delayedAdd") { left: Long, right: Long, delayMs: Long ->
      delayedAdd(left.toULong(), right.toULong(), delayMs.toULong()).toLong()
    }

    Function("startBackgroundTask") {
      startBackgroundTask(object : BackgroundTaskCallback {
        override fun onTick(count: ULong) {
          sendEvent("onTick", mapOf("count" to count.toLong()))
        }
      })
    }

    Function("stopBackgroundTask") {
      stopBackgroundTask()
    }
  }
}
